### PR TITLE
ci(all): in continuous jobs test affected submodules only

### PR DESCRIPTION
**Pull request by nicoleczhu**
Next step: test this PR branch in a temporary job created here: [cl/350847930](https://critique-ng.corp.google.com/cl/350847930)

Fixes: #3454 and #3134
I suspect there could be more edge cases in this CI change. Feedback please! 😄 
_Here's a [one pager summary](https://docs.google.com/document/d/1ZWLb85kkjcnHwUOqnJCexEx_GNf-UVSEAbdfL8A5NGo/edit?resourcekey=0-LO2alh3JcP1DEWFJLshMUA#heading=h.tdx2ho1d7dsd) in case my approach is confusing._

**Changes**
Job | Current | Changes
-- | -- | --
Presubmit | Runs --short tests only<br /> Tests all submodules | No change
Continuous | Runs on all submodules | Runs only the tests in submodules changed by the PR<br /> Runs the root test
Nightly | None | Runs on all submodules<br /> [Optional] can configure a job `nightly/module_name` to only run the tests inside that module, skipping the root test.

- The new `nightly` [job is here](https://fusion.corp.google.com/projectanalysis/summary/KOKORO/prod:cloud-devrel%2Fclient-libraries%2Fgo%2Fgoogle-cloud-go%2Fnightly%2Fmaster?search_pattern=google-cloud-go&include_inactive_projects=false), s.t. all submodules tests still run during the nightly check.

**Open questions:** 
- What are other directories in this repo that, upon being changed in a PR, should trigger CI on ALL submodules? There's `internal` but what else? longrunning?
- Should we keep the `perf-bench` tests e.g. [pubsub](https://fusion.corp.google.com/projectanalysis/summary/KOKORO/prod:cloud-devrel%2Fclient-libraries%2Fgo%2Fgoogle-cloud-go%2Fcontinuous%2Fpubsub_perf_bench?search_pattern=google-cloud-go&include_inactive_projects=false) within the continuous job configs, or should I move them over to nightly? We'll save a bit of compute as these tests are already scheduled to run nightly.

### CODE REVIEW

File: internal/kokoro/continuous.sh [hunk 1]
```diff
@@ -53,23 +54,88 @@ try3 go mod download
...
+GO_CHANGED_MODULES=${GO_CHANGED_MODULES:-./go.mod}
+# Exclude the root module, if present, from the list of sub-modules.
+GO_CHANGED_SUBMODULES=${GO_CHANGED_MODULES#./go.mod}
+
+# Override to determine if all go tests should be run.
+# Does not include static analysis checks.
+RUN_ALL_TESTS="0"
+# If this is a nightly test (not a PR), run all tests.
+if [ -z "${KOKORO_GITHUB_PULL_REQUEST_NUMBER:-}" ]; then
+  RUN_ALL_TESTS="0"
```
Comment (codyoss) [inline internal/kokoro/continuous.sh:79]: This should be true I think?

File: internal/kokoro/continuous.sh [hunk 2]
```diff
@@ -53,23 +54,88 @@ try3 go mod download
...
+CHANGED_DIRS=$(echo "$SIGNIFICANT_CHANGES" | tr ' ' '\n' | grep "/" | cut -d/ -f1 | sort -u | tr '\n' ' ' | xargs ls -d 2>/dev/null || true)
+
+# List all modules in changed directories.
+# If running on master will collect all modules in the repo, including the root module.
+# shellcheck disable=SC2086
+GO_CHANGED_MODULES="$(find ${CHANGED_DIRS:-.} -name go.mod)"
+# If we didn't find any modules, use the root module.
+GO_CHANGED_MODULES=${GO_CHANGED_MODULES:-./go.mod}
+# Exclude the root module, if present, from the list of sub-modules.
+GO_CHANGED_SUBMODULES=${GO_CHANGED_MODULES#./go.mod}
```
Comment (codyoss) [inline internal/kokoro/continuous.sh:72]: This does not look like it is used?

File: internal/kokoro/continuous.sh [hunk 3]
```diff
@@ -53,23 +54,87 @@ try3 go mod download
...
+  runTests .
+else
+  runTests . # Always run root tests.
+  echo "Running tests in modified directories: $CHANGED_DIRS"
+  for d in $CHANGED_DIRS; do
+    mods=$(find "$d" -name go.mod)
+    # If there are no modules, just run the tests directly.
+    if [[ -z "$mods" ]]; then
+      pushd "$d" > /dev/null;
+        runTests
```
Comment (codyoss) [inline internal/kokoro/continuous.sh:120]: If we have changes in `spanner` and `spanner/spansql` this will run tests for all of spanner twice, right? Do we need to dedupe modules with the changed dirs?

File: internal/kokoro/continuous.sh (hunk 1 above)
Comment (nicoleczhu) [reply internal/kokoro/continuous.sh:79]: Good catch thank you! 
I changed it when testing locally. Let me draft this PR as this is a work in progress.

File: internal/kokoro/continuous.sh (hunk 2 above)
Comment (nicoleczhu) [reply internal/kokoro/continuous.sh]: I rewrote the script and scrapped this part

File: internal/kokoro/continuous.sh (hunk 3 above)
Comment (nicoleczhu) [reply internal/kokoro/continuous.sh]: I rewrote the script and scrapped this part

**Comment by nicoleczhu** [comment]
@codyoss Thanks for the previous comments!

I ended up rewriting the build script and created a [CL here](https://critique-ng.corp.google.com/cl/350847930), so I can test the new CI logic using this branch before submitting this PR, just to be safe.

**Comment by nicoleczhu** [comment]
Closing this PR as I repurposed this branch to test this PR's build script.
